Doc-646 TLS revision (#1296)
RedisLabs/redislabs-docs: content/rs/security/tls-ssl.md

categories: ["RS"]
aliases: ["/rs/administering/designing-production/security/tls-configuration", "/rs/administering/designing-production/security/client-connections"]
---
Transport Layer Security (TLS), a successor to SSL, ensures the privacy of data sent between applications and their Redis databases. TLS also secures connections between Redis Enterprise Software nodes.

You can use TLS authentication for one or more of the following types of communication:

- Communication from clients (applications) to your database

> - Communication from your database to other clusters for replication using Replica Of

Reviewer: Maybe hyperlink "Replica Of"

> - Communication to and from your database to other clusters for synchronization using Active-Active

Reviewer: Hyperlink "Active-Active"?

## Enable TLS for client connections

You can enable TLS by editing the configuration of an existing database (as shown below) or by selecting **Advanced Options** when you are creating a new database.

1. Select your database from your database list and navigate to the **configuration** tab.
1. Select **Edit** at the bottom of your screen.
1. Enable **TLS**.

>     - **Enforce client authentication** is selected by default. If you choose to change this option, you will enforce encryption without authentication.

Reviewer: ```suggestion
    - **Enforce client authentication** is selected by default. If you unselect this option, you will still enforce encryption, but TLS client authentication will be disabled.
```

Reviewer: I think it's really important to be clear when talking about TLS client authentication. "Authentication" by itself is ambiguous, as it may refer to password authentication.

1. Save the certificate. ![icon_save](/images/rs/icon_save.png#no-click "Save")
1. Repeat for each client certificate you need to add.
    - If your database uses Replica Of or Active-Active replication, you will need to add the syncer certificates for the participating clusters. The steps for each are below.
1. Optional: To limit connections further to a subset of those with valid certificates, enforce **Subject Alternative Name** and enter authorized users separated with commas.
1. Select **Update** at the bottom of the screen to save your configuration.

## Enable TLS for Active-Active cluster connections

{{< note >}}
You cannot enable or disable TLS after the Active-Active database is created, but you can change the TLS configuration.
{{< /note >}}

### Retrieve syncer certificates

1. For each participating cluster, copy the syncer certificate from the **general** settings tab.

### Configure TLS certificates for Active-Active

1. During database creation (see [Create an Active-Active Geo-Replicated Database]({{<relref "content/rs/administering/creating-databases/create-active-active.md" >}}), select **Edit** from the **configuration** tab.
1. Enable **TLS**.

>     - **Enforce client authentication** is selected by default. If you choose to change this option, you will enforce encryption without authentication.

Reviewer: Same comment as above.

>     - **Enforce client authentication** is selected by default. If you choose to change this option, you will enforce encryption without authentication.

Author: ```suggestion
    - **Enforce client authentication** is selected by default. If you unselect this option, you will still enforce encryption, but TLS client authentication will be disabled.
```

1. Select **Require TLS for CRDB communication only** from the dropdown menu.
    ![crdb-tls-all](/images/rs/crdb-tls-all.png "crdb-tls-all")
1. Select **Add** ![Add](/images/rs/icon_add.png#no-click "Add")

> 1. Paste a syncer certificate into the text box.

Reviewer: ```suggestion
1. Paste certificate or CA to be used for the syncer into the text box.
```

Reviewer: Is this more accurate?

Reviewer 2: It is 👍

Author: I didn't realize the CA feature extended to syncer certificates? I thought it was only for client certificates? @AlonMagrafta can you verify?

Author: Using a CA for the syncer is possible, but complicated and needs more consideration in order to document. DOC-647 will track this work.

    ![database-tls-replica-certs](/images/rs/database-tls-replica-certs.png "Database TLS Configuration")

> 1. Save the syncer certificate. ![icon_save](/images/rs/icon_save.png#no-click "Save")

Reviewer: Same comment here. The "syncer certification" almost sound like a type of certificate. But it's really just a certificate to be used for the syncer process, right?

Reviewer 2: Correct 👍

1. Repeat this process, adding the syncer certificate for each participating cluster.

> 1. Optional: If also you want to  require TLS for client connections, select **Require TLS for All Communications** from the dropdown and add client certificates as well.

Reviewer: ```suggestion
1. Optional: If also you want to require TLS for client connections, select **Require TLS for All Communications** from the dropdown and add client certificates as well.
```

1. Select **Update** at the bottom of the screen to save your configuration.

### Configure TLS on all participating clusters

Repeat this process on all participating clusters.

> To enforce TLS authentication, Active-Active databases require syncer certificates for every cluster connection. If every cluster does not have the syncer certificate for every other participating cluster, it will cause read and write errors.

Reviewer: ```suggestion
To enforce TLS authentication, Active-Active databases require syncer certificates for every cluster connection. If every participating cluster does not have a syncer certificate for every other participating cluster, read and write errors will occur.
```

Reviewer 2: I would suggest: "If every participating cluster does not have a syncer certificate for every other participating cluster, synchronization will fail"
Or something similar. What will happen is the clusters will not even be able to establish the (tls) connection. It will prevent synchronization between the two (or more) clusters w/o the certs, but quickly also affect the "whole" AA DB since it's "waiting" for the isolated replica to communicate.
I hope that was clear

> To enforce TLS authentication, Active-Active databases require syncer certificates for every cluster connection. If every cluster does not have the syncer certificate for every other participating cluster, it will cause read and write errors.

Author: ```suggestion
To enforce TLS authentication, Active-Active databases require syncer certificates for each cluster connection. If every participating cluster doesn't have a syncer certificate for every other participating cluster, synchronization will fail.
```

## Enable TLS for Replica Of cluster connections

You can enable TLS by editing the configuration of an existing database (as shown below) or by selecting **Advanced Options** when you are creating a new database.

> 1. For each cluster hosting a replica, copy the syncer certificate from the **general** settings tab.

Reviewer: ```suggestion
1. For each cluster hosting a replica database, copy the syncer certificate from the **general** settings tab.
```

Author: Alon and I talked about this yesterday. Technically it's not a replica database, it's a replica shard.

Reviewer: Ah. Okay!

1. Select your database from your database list and navigate to the **configuration** tab.
1. Select **Edit** at the bottom of your screen.
1. Enable **TLS**.

>     - **Enforce client authentication** is selected by default. If you choose to change this option, you will enforce encryption without authentication.

Author: ```suggestion
    - **Enforce client authentication** is selected by default. If you unselect this option, you will still enforce encryption, but TLS client authentication will be disabled.
```
